tee: add support for session's client UUID generation

vesajaaskelainen opened the pull request:
TEE Client API defines that from user space only information needed for
specified login operations is group identifier for group based logins.

REE kernel is expected to formulate trustworthy client UUID and pass that
to TEE environment. REE kernel is required to verify that provided group
identifier for group based logins is verified against calling processes
group memberships.

Linux kernel does not provide common contex for application identifier,
instead different security frameworks provide own means to defines
application identifier for running process. Code includes place holder for
such solutions but is left for later implementation.

TEE specification only defines that the information passed from REE
environment to TEE environment is encoded into on UUID.

In order to guarantee trustworthiness of client UUID user space is not
allowed to freely pass client UUID.

UUIDv5 form is used encode variable amount of information needed for
different login types.

Signed-off-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>

This is related to: https://github.com/OP-TEE/optee_os/issues/3642

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
+	}
+
+	/*
+	 * In Linux environment client UUID is based on UUIDv5.
```
Do you have a link/reference to some standard or convention?

vesajaaskelainen replied:
Probably the easiest is to have a look at wikipedia:
https://en.wikipedia.org/wiki/Universally_unique_identifier

And then the actual RFC:
https://www.rfc-editor.org/rfc/rfc4122.txt

If you talk about the "Linux environment" part this is an agreement that needs to be made. GlobalPlatform leaves this as undefined and it is up to REE implementation to decide how to do it.

But in the end it must be UUID and only deterministic choices are then UUIDv3 or UUIDv5 where UUIDv5 is slightly better.

jenswikl replied:
Yeah, it was rather that GP leaves this undefined and we're improvising or proposing a way to deal with it. Boot loaders may access TAs too, so we need something that can be replicated there if needed.

vesajaaskelainen replied:
In this case I would say there needs to be unique namespace UUID allocated for that case. Perhaps this should be managed in OP-TEE as a registry or so.

jenswikl replied:
> Perhaps this should be managed in OP-TEE as a registry or so.

If needed later on we'll do something about that. What's important here is to be clear that GP leaves this undefined and we're filling the gap in this way.

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
 	return 0;
 }
 
+static const char *tee_session_get_application_id(void)
```
These two functions seems rather useless. Until they do anything useful how about just skipping them?

vesajaaskelainen replied:
idea is that in here we would have in example SMACK or SELnux labels for particular application. Eg. get the application label from the context. but that code I do not have at the moment.

I believe this should be some kernel method to configure source for application id and then per system this is selected. Eg. if SMACK is in use then smack application label from current process would be used. Some may have non-upstreamed ways to deliver application ID and then they could provide with this mechanism.

Idea was here to illustrate that something like this is needed.

jenswikl replied:
> Idea was here to illustrate that something like this is needed.

OK, that's good for discussion. But if it's to go into the kernel we either need something specific and usable or have it removed.

vesajaaskelainen replied:
I was also thinking an option to split this commit as two.

One which does user and group and then leaves application part to own commit as RFC.

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
+	struct shash_desc *desc = NULL;
+	int rc;
+
+	shash = crypto_alloc_shash("sha1", 0, 0);
```
Any particular reason why we're using SHA-1?

vesajaaskelainen replied:
UUIDv5 is SHA1 based. Much better than UUIDv3 which is MD5 based.

jenswikl replied:
I see. A comment explaining that we're hashing for UUIDv5 as described in RFCxxxx would be useful.

vesajaaskelainen replied:
method is called as uuid_v5 :) ? I can add comment about RFC too there.

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
+
+	case TEE_IOCTL_LOGIN_APPLICATION:
+		application_id = tee_session_get_application_id();
+		scnprintf(name, TEE_UUID_NS_NAME_SIZE, "app=%s",
```
This will currently result in the string `app=(null)` or similar. I'm not sure this is the best. How about just hardcode it to "app=" until we know what to do with it?

vesajaaskelainen replied:
Or return error if application ID is null?

jenswikl replied:
Yes, that could be an option. We need to be careful to not break compliance testing though.

vesajaaskelainen replied:
So far the client id handling in kernel is broken... Eg. I would be very surprised that it worked at all for any compliance testing so I do not see that as a worry.

jenswikl replied:
Given that the resulting UUID passed to the TA is not specified the GP tests may very well be satisfied with not returning an error.

vesajaaskelainen replied:
Probably safest would be to return -EINVAL or so if app id is NULL so that there would not be safe feeling that everything is working correctly.

jenswikl commented on include/linux/tee_drv.h:
```diff
@@ -169,6 +169,32 @@ int tee_device_register(struct tee_device *teedev);
+/**
+ * tee_session_calc_client_uuid() - Calculates client UUID for session
+ * @uuid:		Resulting UUID
+ * @connection_method:	Connection method for session
```
Could be worth mentioning that accepted values are from `TEE_IOCTL_LOGIN_*`

vesajaaskelainen replied:
ok

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
+	if (unlikely(rc < 0))
+		goto out2;
+
+	// time_low
```
How about something shorter like:
``` C
memcpy(uuid->b, hash, TEE_IOCTL_UUID_LEN);
/* Tag it UUID V5 */
uuid->b[6] = (uuid->b[6] & 0x0F) | 0x50;
uuid->b[8] = (uuid->b[8] & 0x3F) | 0x80;

vesajaaskelainen replied:
Can do that too. There is just then some extra copies. Compiler might be actually pretty smart here and optimize this quite nicely automatically.

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
+		goto out2;
+
+	rc = crypto_shash_update(desc, (const u8 *)ns, sizeof(*ns));
+	if (unlikely(rc < 0))
```
I don't think much is gained with `unlikely()` here.

vesajaaskelainen replied:
Ok.

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -220,6 +220,16 @@ int optee_open_session(struct tee_context *ctx,
 	struct optee_session *sess = NULL;
+	uuid_t client_uuid;
+
+	rc = tee_session_verify_perm(arg->clnt_login, arg->clnt_uuid);
```
Since this is how these two functions are supposed to be used, how about having only one that does it all?
Perhaps we could put the result directly at `&msg_arg->params[1].u.value` in order to save some stack?

vesajaaskelainen replied:
Of course this is now about OP-TEE path here but there is also a AMD path in kernel so I am wondering if that would need the same treatment and in there it might be different.

Basic idea was to keep policy and generation as a separate.

jenswikl replied:
Right now a single function would be best, so let's have that.
Please move the OP-TEE changes into a separate patch.

vesajaaskelainen replied:
ok

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -6,18 +6,28 @@
 
+#define TEE_UUID_NS_NAME_SIZE	128
+
+/* TEE Client UUID name space identifier (UUIDv4) */
```
How does this relate to UUIDv5? From where did you get these numbers?

vesajaaskelainen replied:
UUIDv5 needs to have UUID namespace identifier.

UUIDv4 is random and it was generated by my Linux kernel random number generation with uuidgen tool.

jenswikl replied:
So it's the namespace identifier for this UUID scheme, that is, the part left implementation defined by GP. That needs to be explained in a comment.

vesajaaskelainen replied:
Ok. will try to improve the wording.

vesajaaskelainen committed:
[RFC] tee: add support for app id for client UUID generation

Linux kernel does not provide common contex for application identifier,
instead different security frameworks provide own means to defines
application identifier for running process. Code includes place holder for
such solutions but is left for later implementation.

Open questions:

1. App ID source

How to specify what source is used for app id?

Does it need to be protected on runtime?
- Should this be Kconfig setting?
- Cnfigure once during runtime thru sysfs or so?
- Configure from device tree?

2. Formatting for App ID

Should there be common format? Or common keyword id?

3. How to handle custom App ID sources

Android has own App ID so does Tizen.

Should there be place holder for this where to make local patch?

Signed-off-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>

vesajaaskelainen commented:
@jenswi-linaro I believe I have now addressed your comments and updated the PR accordingly.

I have moved application ID part as own RFC commit that could easily left out from initial submissions as that is open and could take longer period of time to have a solution.

For convenience of review I have created -delta branch to see individual edits:
https://github.com/vesajaaskelainen/linux/commits/linaro-optee-login-checks-delta

jenswikl commented on drivers/tee/tee_core.c:
```diff
@@ -111,6 +126,162 @@ static int tee_release(struct inode *inode, struct file *filp)
 }
 
+static const char *tee_session_get_application_id(void)
+{
```
These two functions are unused in this patch, should be moved into the "RFC" patch.

vesajaaskelainen replied:
Ash... they were there at one point of time. will fix it. Sorry about it.

jenswi-linaro reviewed (commented):
Good thinking with the moving parts into an RFC patch.
Some comments on "drivers: tee: add support for session's client UUID generation"

vesajaaskelainen committed:
tee: add support for session's client UUID generation

TEE Client API defines that from user space only information needed for
specified login operations is group identifier for group based logins.

REE kernel is expected to formulate trustworthy client UUID and pass that
to TEE environment. REE kernel is required to verify that provided group
identifier for group based logins matches calling processes group
memberships.

TEE specification only defines that the information passed from REE
environment to TEE environment is encoded into on UUID.

In order to guarantee trustworthiness of client UUID user space is not
allowed to freely pass client UUID.

UUIDv5 form is used encode variable amount of information needed for
different login types.

Signed-off-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>

vesajaaskelainen committed:
tee: optee: Add support for session login client UUID generation

Adds support for client UUID generation for OP-TEE. For group based session
logins membership is verified.

Signed-off-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>

vesajaaskelainen commented:
@jenswi-linaro Now it should be better and splitted.

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -220,6 +220,12 @@ int optee_open_session(struct tee_context *ctx,
 	struct optee_session *sess = NULL;
+	uuid_t client_uuid;
+
+	rc = tee_session_calc_client_uuid(&client_uuid, arg->clnt_login,
```
This call can be moved to replace the `memcpy()` below and saving us from keeping `client_uuid` on the stack.
The reason I care is the call to `optee_do_call_with_arg()` which may do some rather deep function calls during RPC handling so `client_uuid` will directly add to the maximum amount of stack used. Since this saving doesn't cost anything I can see no reason to not do it.

vesajaaskelainen replied:
ok

jenswi-linaro reviewed (commented):
Looks good. You can remove the `drivers: ` prefix from the first line of the commit messages, `tee: ` is an established prefix in itself.

vesajaaskelainen commented:
@jenswi-linaro updated. Removed "drivers: " from commit messages and moved Client UUID permission check and generation later.

Now there is ugly cast but I suppose it serves the purpose and I suppose the alignment is handled already.

jenswi-linaro commented:
With this we can merge "tee: add support for session's client UUID generation" and "tee: optee: Add support for session login client UUID generation" to linaro-swg with the purpose of demonstrating the added test cases.

When it comes to "[RFC] drivers: tee: add support for app id for client UUID generation". I don't think that upstream is much interested in providing placeholders for out of tree changes. The aim should be to in the end have something that makes sense by itself upstream.

For these patches to land upstream I'd prefer if you can post these patches on the mailing lists, don't forget to add me and the AMD guys and anyone else you'd like to get the attention of in CC. If all goes well I'll pick up the first two patches for upstream. With the RFC patch we'll see what comes out of the discussion.

@vesajaaskelainen, does this make sense to you?

vesajaaskelainen commented:
> With this we can merge "tee: add support for session's client UUID generation" and "tee: optee: Add support for session login client UUID generation" to linaro-swg with the purpose of demonstrating the added test cases.
> 
> When it comes to "[RFC] drivers: tee: add support for app id for client UUID generation". I don't think that upstream is much interested in providing placeholders for out of tree changes. The aim should be to in the end have something that makes sense by itself upstream.
> 
> For these patches to land upstream I'd prefer if you can post these patches on the mailing lists, don't forget to add me and the AMD guys and anyone else you'd like to get the attention of in CC. If all goes well I'll pick up the first two patches for upstream. With the RFC patch we'll see what comes out of the discussion.
> 
> @vesajaaskelainen, does this make sense to you?

I agree that the RFC part is probably going to need much of discussion.

I think that merging two first would get us forward and then later the RFC part.

I can send the emails later this week. Need to rebase non-linaro branch to latest kernel version and test it out.

Do you want me to drop the RFC commit from this PR?

jenswi-linaro commented:
> Do you want me to drop the RFC commit from this PR?

It would be nice to keep it for reference. Perhaps @jforissier can take the first two commits and then close this PR manually.

jforissier commented:
> > Do you want me to drop the RFC commit from this PR?
> 
> It would be nice to keep it for reference. Perhaps @jforissier can take the first two commits and then close this PR manually.

No problem.

jforissier commented:
@jenswi-linaro do you want to provide some tags for the first two commits?

jenswi-linaro commented:
I usually avoid doing that in order to avoid confusion with patches posted on the mailing lists.

Now I realize that I just did that in https://github.com/linaro-swg/linux/pull/72, but that PR is still far away from upstream and I didn't want to have to look through those patches again and again.

jforissier commented:
> I usually avoid doing that in order to avoid confusion with patches posted on the mailing lists.

OK, both ways are OK with me. I'm picking up the two patches and applying to our optee branch.
